File: PreviewDocumentView.m
#import "PreviewDocumentView.h"
#import "PreviewLayer.h"
#import <QuartzCore/QuartzCore.h>

@implementation PreviewDocumentView {
    NSString *_mixerId;
    PreviewLayer *_layer;
}

- (instancetype)initWithFrame:(NSRect)frameRect
{
    self = [super initWithFrame:frameRect];
    if (self) {
        _layer = [PreviewLayer new];

        self.layer = _layer;
        self.wantsLayer = TRUE;
    }
    return self;
}

- (NSString *)mixerId
{
    return _mixerId;
}

- (void)setMixerId:(NSString *)mixerId
{
    if ([mixerId isEqualToString:_mixerId])
        return;

    _mixerId = mixerId;
    [self updatePreviewMixerId];
}

- (void)viewDidMoveToSuperview
{
    self.needsLayout = TRUE;
    [self updatePreviewMixerId];
}

- (void)resizeWithOldSuperviewSize:(NSSize)oldSize
{
    self.needsLayout = TRUE;
}

- (void)updatePreviewMixerId
{
    _layer.mixerId = self.superview != nil ? _mixerId : nil;
}

- (void)layout
{
    CGSize size = self.superview.bounds.size;
    CGRect frame = CGRectMake(0, 0, size.width, size.height);

    self.frame = frame;
    _layer.frame = frame;

    [super layout];
}

- (void)setDataSource:(WebDataSource *)dataSource
{
}

- (void)dataSourceUpdated:(WebDataSource *)dataSource
{
    NSString *str = [[NSString alloc] initWithData:dataSource.data encoding:NSUTF8StringEncoding];
    if (str)
        self.mixerId = str;
}

- (void)viewWillMoveToHostWindow:(NSWindow *)hostWindow
{
}

- (void)viewDidMoveToHostWindow
{
}

@end

File: PreviewDocumentView.swift
import Cocoa
import WebKit

class PreviewDocumentView: NSView, WebDocumentView {

    private let previewLayer = PreviewLayer()

    var mixerId: String? {
        didSet {
            updatePreviewMixerId();
        }
    }

    required init?(coder: NSCoder) {
        super.init(coder: coder);
        setupPreviewLayer();
    }

    override init() {
        super.init();
        setupPreviewLayer();
    }

    override init(frame frameRect: NSRect) {
        super.init(frame: frameRect);
        setupPreviewLayer();
    }

    func setupPreviewLayer() {
        layer = previewLayer;
        wantsLayer = true;
    }

    override func viewDidMoveToSuperview() {
        updatePreviewMixerId();
        needsLayout = true;
    }

    override func resizeWithOldSuperviewSize(oldSize: NSSize) {
        needsLayout = true;
    }

    func updatePreviewMixerId() {
        previewLayer.mixerId = superview != nil ? mixerId : nil;
    }

    func setNeedsLayout(flag: Bool) {
        super.needsLayout = flag;
    }

    override func layout() {
        let size = self.superview!.bounds.size;
        let newFrame = NSMakeRect(0, 0, size.width, size.height);

        frame = newFrame;
        previewLayer.frame = newFrame;

        super.layout();
    }

    func setDataSource(dataSource: WebDataSource!) {
    }

    func dataSourceUpdated(dataSource: WebDataSource!) {
        mixerId = NSString(data: dataSource.data, encoding: NSUTF8StringEncoding);
    }

    func viewWillMoveToHostWindow(hostWindow: NSWindow!) {
    }

    func viewDidMoveToHostWindow() {
    }

}
